main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala

```diff
@@ -394,6 +399,26 @@ object Monomorph {
         }
         Expression.Match(visitExp(exp, env0), rs, subst0(tpe), pur, eff, loc)
 
+      case Expression.MatchType(exp, rules, tpe, pur, eff, loc) =>
+        // make the tvars in `exp`'s type rigid
+        // so that Nil: List[x%123] can only match List[_]
+        val renv = exp.tpe.typeVars.foldLeft(RigidityEnv.empty) {
+          case (acc, Type.Var(sym, _)) => acc.markRigid(sym)
+        }
+        rules.collectFirst {
+          case MatchTypeRule(sym, t, body0)
+            // use the non-strict substitution
+            // to allow free type variables to match with anything
+            if Unification.unifiesWith(subst0.nonStrict(exp.tpe), subst0.nonStrict(t), renv) =>
```

Comment:
This does a check, but it does not compute a new substitution. What if I have:

```
typematch "abc" :: Nil {
  case x: List[a] => let y: a = head(x)
}
```

then `y` is never given the right type?

Reply:
My intention is not to introduce new type variables here. In your example you'd get an error saying `no such variable "a"`.

I think this is the right choice because by ensuring the type variables are previously bound, we can do things like:

```scala
def sameType(x: a, y: b): Bool = typematch x {
    case _: b => true
    case _: _ => false
}
```

Reply:
Either `a` is not in the signature -- résolution error -- or it is in the signature and has been monomorphed to something so it actually looks like `case x: List[MyCoolEnum]`